crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -387,6 +383,82 @@ impl QueryBuilder {
+        filter: Option<EventFilter>,
+    ) -> Result<events::BoxedQuery<'a, Pg>, Error> {
+        let mut query = events::dsl::events.into_boxed();
+        if let Some(cursor_val) = cursor {
+            if descending_order {
+                query = query.filter(events::dsl::tx_sequence_number.lt(cursor_val.0));
+                query = query.filter(events::dsl::event_sequence_number.lt(cursor_val.1));
+            } else {
+                query = query.filter(events::dsl::tx_sequence_number.gt(cursor_val.0));
+                query = query.filter(events::dsl::event_sequence_number.gt(cursor_val.1));
+            }
+        }
+        if descending_order {
+            query = query.order((
+                events::dsl::tx_sequence_number.desc(),
+                events::dsl::event_sequence_number.desc(),
+            ));
+        } else {
+            query = query.order((
+                events::dsl::tx_sequence_number.asc(),
+                events::dsl::event_sequence_number.asc(),
+            ));
+        }
+        query = query.limit(limit + 1);
+
+        if let Some(filter) = filter {
+            if let Some(sender) = filter.sender {
+                let subquery = tx_senders::dsl::tx_senders
+                    .filter(tx_senders::dsl::sender.eq(sender.into_vec()))
+                    .select(tx_senders::dsl::tx_sequence_number);
+
+                query = query.filter(events::dsl::tx_sequence_number.eq_any(subquery));
+            }
+
+            if let Some(digest) = filter.transaction_digest {
+                let tx_digest = Digest::from_str(&digest)?.into_vec();
+                let subquery = transactions::dsl::transactions
+                    .filter(transactions::dsl::transaction_digest.eq(tx_digest))
+                    .select(transactions::dsl::tx_sequence_number);
+
+                query = query.filter(events::dsl::tx_sequence_number.eq_any(subquery));
+            }
+
+            match (filter.emitting_package, filter.emitting_module) {
+                (Some(p), None) => {
+                    query = query.filter(events::dsl::package.eq(p.into_vec()));
+                }
+                (Some(p), Some(m)) => {
+                    query = query.filter(events::dsl::package.eq(p.into_vec()));
+                    query = query.filter(events::dsl::module.eq(m));
+                }
+                _ => {}
+            }
+
+            match (filter.event_package, filter.event_module, filter.event_type) {
+                (Some(p), None, None) => {
+                    query = query.filter(events::dsl::event_type.like(format!("{}::%", p)));
+                }
+                (Some(p), Some(m), None) => {
+                    query = query.filter(events::dsl::event_type.like(format!("{}::{}::%", p, m)));
+                }
+                (Some(p), Some(m), Some(t)) => {
+                    query =
+                        query.filter(events::dsl::event_type.like(format!("{}::{}::{}%", p, m, t)));
```

Comment:
This is a compromise to handle both event types that look like 0x3::validator::StakingRequestEvent
vs 0x3164fcf73eb6b41ff3d2129346141bd68469964c2d95a5b1533e8d16e6ea6e13::Market::ChangePriceEvent<0x2::sui::SUI> (type w/ generics)

Reply:
We should be more structured here -- if `t` includes angle brackets, then they should be balanced and it should be a precise match, if not, then we should be prepared for the case where it is either a concrete type without generics (no prefix match) or it's the start of a generic type, in which case it should be like `{}::{}::{}<%`.

Otherwise people will get unexpected results, like they searched for `0xf00::bar::Baz` and they got `0xf00::bar::Bazaar`.

Reply:
Yes thanks for calling that out. When it comes to a client trying to filter by type, we'll run into three kinds of users:
1. User trying to fetch generic type event, knows the inner T
2. User trying to fetch non-generic type
3. User wants to fetch generic type, doesn't know is generic, doesn't know inner T
4. User wants to fetch generic type, doesn't care what the inner T is, just wants all of the type

If a user falls in categories 3 or 4, the service would not know how to make the proper query without looking up the db. I would prefer not to have our service do this additional work (similar to the 2x lookup to resolve dynamic object fields.) What about modifying the filter to have an optional 'exactType' and 'rootType' where the latter is meant for just the start of a generic type like 0x2::a::B of 0x2::a::B<...>

Reply:
My point is that there are the following types of valid type query:

- `package` -- match against all types in the package (and not packages whose prefix is `package`)
- `package::module` match against all types in the module (and not modules whose prefix is `module`)
- `package::module::Type` match against all instances of `Type`, including all type instantiations if it is generic, and not types whose prefix is `Type`.
- `package::module::Type<$T, ... $Ts>` match against all instances of the type instantiation.

We do not need to separate out `exactType` and `rootType`, the filter + context will always imply one, we just need to do a more thorough job with validating it and crafting the query:

- When we are given a type without angle brackets, our query needs to either search for that type with an open angle bracket at the end, or with EOS, e.g.:
 
```sql
 WHERE
   eventType LIKE '{package}::{module}::{name}<%' 
OR eventType = '{package}::{module}::{name}'

```

- When validating the query, we need to make sure that all its parts (package, module, name) are valid move identifiers, and that if the name includes angle brackets, they are well bracketed, and they surround one or more fully qualified types.